### ISSUE TITLE
Score calculations

### ISSUE THREAD

**Issue by okhin** [political-memory/political_memory#63]
An important issue is, I think, to think (and tell) how score are calculated.
I'm not sure the basic total is a good thing. For starters, it doesn't says anything about the current position of a representative about a dossier, or a thematics.It says about the past, but not about how they votes.
Also, a socre of 5 means what? It's a mean score, a good one, a bad one? We need a scale. At least we should now what was the mean score for each votes and compare the representative one to the mean.
Maybe it could be interseting to know the mean score for each entities (political groups, comission, mandates, etc), for instance a representative with a score of 8 on a vote about Net Neutrality could be in line with LIBE for instance (which might have a mean socre of 7.6) and out of line with their group - S&D for instance which could have a mean score of -4.5.

A mean score might be interesting, but it doesn't show how - or if - the position of a representatives evolve over time, which is something important to measure, it helps for feedback of campaining, it helps to detect which of them are having doubts and are easy targets for lobbying, etc.

Maybe using a [Moving Average](https://en.wikipedia.org/wiki/Moving_average), or something similar, which will emphasize the recent votes.

Or maybe adding a trend to the score (getting better, getting worse).

**Comment by jpic** [comment]
At least, we could use days-since-vote as a factor to emphasize on recent
votes, would that work for now ?

**Comment by njoyard** [comment]
Maybe we could have a broader brainstorm about scores altogether (in order to have something flexible on which we can build simple & configuratble computation rules). Il try to draft something in the next few days.

**Comment by stef** [comment]
i think if you find the mailing list archive you can see lot's discussion, brainstorming and reasons for people leaving the development of memopol. ;)

**Comment by Psycojoker** [comment]
13:55  <Bram> btw, regarding score I've made this prototype in the past http://pix.toile-libre.org/upload/original/1390741197.png
13:55  <Bram> hacking some d3js I've found
13:55  <Bram> the bigger the stuff, the most weight this mep has
13:56  <Bram> and you get an idea on which position he is and up to which point he could have been

I think that the code is in some branch on one the repos but I can't remember which one :/

![1390741197](https://cloud.githubusercontent.com/assets/41827/14915331/5bcc3252-0e12-11e6-8816-d80cec31bdbc.png)

**Comment by Psycojoker** [comment]
But yeah, as @stef said, go read the ml, this quesiton has already been discusted a **lot**.

**Comment by njoyard** [comment]
@Psycojoker yeah, already started :)

**Comment by njoyard** [comment]
Depends on #79

**Comment by njoyard** [comment]
About score computation, we could get quite good control over score decay by using a formula like:
```
decayed_score = vote_score * exp( -(vote_age/DECAY_RATE)^2*EXPONENT )
```
With two parameters (which would be easy to implement thanks to the new view-computed score):
* `DECAY_RATE` sets an "age threshold", ie. how long votes stay meaningful; for example setting it to 5 years would say "votes from the current mandate are mostly untouched, older votes decay rapidly"
* `EXPONENT` sets how fast scores decay around the threshold

For example here (X scale is in days) with `DECAY_RATE=1 year` and `EXPONENT=1`, there's a quite gradual decay and votes after 2 years are nearly ignored:
![two](https://framapic.org/7oQazP3XWgSI/KeO52QIgve7B.png)

Increasing the `DECAY_RATE` (10 years here) makes the score stay strong longer ; raising it to infinity makes it stay identical whatever the age (ie. what we have now):
![three](https://framapic.org/GIjm60v3DakG/prn5mHvyCUVW.png)

Instead, increasing `EXPONENT` makes the decay steeper:
![one](https://framapic.org/MSgBtoJ9O9RG/dVPKH4i04OeE.png)

We can even increase it dramatically to set a brutal cutoff threshold:
![four](https://framapic.org/9qaGBmS4AAU1/FR64ZQprYLDh.png)

**Comment by njoyard** [comment]
Note that the point here is to give *control* over score computation, not determine which is the best - thats not a developer issue :D

**Comment by jpic** [comment]
Seriously, @njoyard, :guitar: :guitar: :guitar: :guitar: :guitar: :guitar:

@okhin could you let us know what's the opinion of our friends @ LQDN on this ? Thanks !!

**Comment by njoyard** [comment]
> would a simple django setting do the trick ?

Something stored in database would be better. That way we can include the computation in score views (DB-views).

**Comment by jpic** [comment]
Ok, then perhaps we could add another "Setting" model, usable with the admin ?

Perhaps an app like this would help ?

http://django-constance.readthedocs.io/en/latest/

More to compare:

https://www.djangopackages.com/grids/g/live-setting/

**Comment by okhin** [comment]
Seems a really good idea, and I like the fact that each instance admin can choose how to affect the scores :)

**Comment by njoyard** [comment]
Oops forgot to answer @jpic:
> you're going to implement this is an SQL view right ?
of course :)

**Comment by njoyard** [comment]
Closing this as:
* We implemented configurable decay
* We have a plan in design v3 to improve score displays :
  * Display score trend
  * Display a radial graph with score on each theme, on a [worst, last] actual mep score scale
* We will add manual scores on positions (and score bonuses for vote against group), so that makes "what is the best/worst score the mep could have possibly add" no longer relevant.
